### PR TITLE
Add AdSense linked and last synced at debug fields.

### PR DESCRIPTION
**Pull request by benbowler**
## Summary

Addresses issue:

- #8051 

## Relevant technical choices

I implemented the debug fields as in the IB/AC, however I believe the test in the IB was not correct, the test mentioned covers the `get_datapoints` function rather than `get_debug_fields` to which this ticket refers. Instead, I created a new test `test_get_debug_fields` which follows the patterns of the `test_get_datapoints__news_key_metrics` test to check for expected keys in this functions response.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Modules/Analytics_4.php [hunk 1]
```diff
@@ -367,6 +367,16 @@ public function get_debug_fields() {
 				'value' => $settings['useSnippet'] ? __( 'Yes', 'google-site-kit' ) : __( 'No', 'google-site-kit' ),
 				'debug' => $settings['useSnippet'] ? 'yes' : 'no',
 			),
+			'adsense_linked'                 => array(
+				'label' => __( 'AdSense Linked', 'google-site-kit' ),
+				'value' => $settings['adSenseLinked'] ? __( 'Connected', 'google-site-kit' ) : __( 'Not Connected', 'google-site-kit' ),
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:372]: Very nitpicky, but according to the ACs and other debug entries, I think we don't want a capitalized string for the values.

The labels seem already very inconsistent with the casing, so let's disregard that for now 😅 

```suggestion
				'value' => $settings['adSenseLinked'] ? __( 'Connected', 'google-site-kit' ) : __( 'Not connected', 'google-site-kit' ),
```

File: includes/Modules/Analytics_4.php [hunk 2]
```diff
@@ -367,6 +367,16 @@ public function get_debug_fields() {
...
+				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinked'] ),
+			),
+			'adsense_linked_last_synced_at'  => array(
+				'label' => __( 'AdSense Linked Last Synced At', 'google-site-kit' ),
+				'value' => $settings['adSenseLinkedLastSyncedAt'] ? gmdate( 'Y-m-d H:i:s', $settings['adSenseLinkedLastSyncedAt'] ) : __( 'Never Synced', 'google-site-kit' ),
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:377]:
```suggestion
				'value' => $settings['adSenseLinkedLastSyncedAt'] ? gmdate( 'Y-m-d H:i:s', $settings['adSenseLinkedLastSyncedAt'] ) : __( 'Never synced', 'google-site-kit' ),
```

File: includes/Modules/Analytics_4.php [hunk 3]
```diff
@@ -367,6 +367,16 @@ public function get_debug_fields() {
 				'value' => $settings['useSnippet'] ? __( 'Yes', 'google-site-kit' ) : __( 'No', 'google-site-kit' ),
 				'debug' => $settings['useSnippet'] ? 'yes' : 'no',
 			),
+			'adsense_linked'                 => array(
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:370]: The ACs say that these entries should only be added if both the Analytics & AdSense modules are connected.

> * The `adsenseLinked` setting should be output in the Site Health settings when both Analytics 4 and AdSense modules are connected.

While I think it is not quite easy to see if AdSense is connected as it is outside the scope of this class, we can instantiate the `Modules\AdSense\Settings` class and check if that meets the AdSense connection requirements (see `Modules\AdSense::is_connected`).

@tofumatt Please let us know if this is not a must have.

File: includes/Modules/Analytics_4.php [hunk 4]
```diff
@@ -367,6 +367,16 @@ public function get_debug_fields() {
...
+				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinked'] ),
+			),
+			'adsense_linked_last_synced_at'  => array(
+				'label' => __( 'AdSense Linked Last Synced At', 'google-site-kit' ),
+				'value' => $settings['adSenseLinkedLastSyncedAt'] ? gmdate( 'Y-m-d H:i:s', $settings['adSenseLinkedLastSyncedAt'] ) : __( 'Never Synced', 'google-site-kit' ),
+				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinkedLastSyncedAt'] ),
+			),
 		);
 
 		$debug_fields['analytics_4_available_custom_dimensions'] = array(
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:382]: This was originally added as a separate entry to the `$debug_fields` array as it was conditional based on the availability of a feature flag. This ideally should've been added to the original array when the feature flag was removed.

For the sake of cleanliness, would you be able to move it above as a part of the original array? Thanks!

**Review by nfmohit — CHANGES_REQUESTED** [review]
Hi @benbowler, brilliant work here, thanks! I have left some minor nitpicky comments. Please let me know what you think about them. Thanks!

File: includes/Modules/Analytics_4.php (hunk 3 above)
Comment (benbowler) [reply includes/Modules/Analytics_4.php]: @nfmohit I have implemented a solution as mentioned here, although I have not been able to connect the AdSense module on my local for testing, if this is not the correct solution, I suggest we jump on a short call and test/implement together.

https://github.com/google/site-kit-wp/blob/8339d042c5f1ce0127f825530d6a009ab0ed9203/includes/Modules/Analytics_4.php#L335-L340

File: includes/Modules/Analytics_4.php [hunk 5]
```diff
@@ -331,6 +332,13 @@ public function on_deactivation() {
 	public function get_debug_fields() {
 		$settings = $this->get_settings()->get();
 
+		$adsense_connected = true;
+		$adsense_settings  = ( new AdSense_Settings( $this->options ) )->get();
+
+		if ( empty( $adsense_settings['accountSetupComplete'] ) || empty( $adsense_settings['siteSetupComplete'] ) ) {
+			$adsense_connected = false;
+		}
+
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:341]: Let's move this to its own method that returns a boolean, e.g. `is_adsense_connected`. What do you think?

File: includes/Modules/Analytics_4.php [hunk 6]
```diff
@@ -347,40 +355,49 @@ public function get_debug_fields() {
...
+				'label' => __( 'AdSense Linked', 'google-site-kit' ),
+				'value' => $adsense_connected && $settings['adSenseLinked'] ? __( 'Connected', 'google-site-kit' ) : __( 'Not connected', 'google-site-kit' ),
+				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinked'] ),
+			),
+			'adsense_linked_last_synced_at'           => array(
+				'label' => __( 'AdSense Linked Last Synced At', 'google-site-kit' ),
+				'value' => $adsense_connected && $settings['adSenseLinkedLastSyncedAt'] ? gmdate( 'Y-m-d H:i:s', $settings['adSenseLinkedLastSyncedAt'] ) : __( 'Never synced', 'google-site-kit' ),
+				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinkedLastSyncedAt'] ),
+			),
 		);
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:401]: Let's add these two entries to the array only if both Analytics & AdSense are connected.

```suggestion
		);
		
		if ( $this->is_adsense_connected() ) {
			$debug_fields['adsense_linked'] = array(
				'label' => __( 'AdSense Linked', 'google-site-kit' ),
				'value' => $settings['adSenseLinked'] ? __( 'Connected', 'google-site-kit' ) : __( 'Not connected', 'google-site-kit' ),
				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinked'] ),
			);

			$debug_fields['adsense_linked_last_synced_at'] = array(
				'label' => __( 'AdSense Linked Last Synced At', 'google-site-kit' ),
				'value' => $settings['adSenseLinkedLastSyncedAt'] ? gmdate( 'Y-m-d H:i:s', $settings['adSenseLinkedLastSyncedAt'] ) : __( 'Never synced', 'google-site-kit' ),
				'debug' => Debug_Data::redact_debug_value( $settings['adSenseLinkedLastSyncedAt'] ),
			);
		}
```

**Review by nfmohit — CHANGES_REQUESTED** [review]
Hi @benbowler. Back to you with one more minor feedback. Thanks!

File: includes/Modules/Analytics_4.php (hunk 3 above)
Comment (nfmohit) [reply includes/Modules/Analytics_4.php]: This looks good to me, but left a small follow-up feedback. Thanks!

**Comment by benbowler** [comment]
@nfmohit great suggestions, I've updated the PR and back to you in CR.

**Comment by benbowler** [comment]
OK @nfmohit, back into CR with suggested changes plus and additional test case.

File: includes/Modules/Analytics_4.php [hunk 7]
```diff
@@ -321,6 +322,20 @@ public function on_deactivation() {
 		$this->custom_dimensions_data_available->reset_data_available();
 	}
 
+	/**
+	 * Checks whether the AdSense module is connected.
+	 *
+	 * @since n.e.x.t
+	 */
```
Comment (nfmohit) [inline includes/Modules/Analytics_4.php:329]: I believe we want to include a return statement here.

**Review by nfmohit — CHANGES_REQUESTED** [review]
Nice work @benbowler. One last small change, so small that I applied that myself. I hope that's okay!

**Review by nfmohit — APPROVED** [review]
LGTM, thank you @benbowler!

**Review by tofumatt — APPROVED** [review]

**Pull request merged**